### PR TITLE
Editing via esc does not correctly update node view

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

Fixes https://www.pivotaltracker.com/story/show/182926584
[Task link](https://www.pivotaltracker.com/story/show/183426449)

This PR fixes an IDE freeze introduced by https://github.com/enso-org/enso/pull/3732 and reimplements reverting edited nodes to their previous state. 

The cause of the IDE freeze is quite interesting. A detailed investigation is available [here](https://gist.github.com/vitvakatu/785e34881368b8cfda61715d7543cbd0).

The graph editor needs to update the Presenter state only if the user is editing the node. Before this PR, the graph editor notified the Presenter with a visual representation of the node content instead of code expression. It caused inconsistency between the states of the controller and Presenter and caused severe performance issues.

https://user-images.githubusercontent.com/6566674/195831224-6d6e8258-e347-48b4-890a-d89c7300bc39.mp4

### Important Notes

- ~~There is a more complex alternative solution – it requires refactoring of the `component::node::input::area` module. The Presenter can be notified with `expression.code` changes, not `expression.viz_code`. I found a simpler solution (`.gate(&edit_mode)`), which has the same effect but does not require additional refactoring.~~ Said solution is implemented in a separate commit

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/gui/src/presenter/graph/state.rs [hunk 1]
```diff
@@ -659,18 +659,24 @@ impl<'a> ViewChange<'a> {
     }
 
     /// Set the node expression.
-    pub fn set_node_expression(&self, id: ViewNodeId, expression: String) -> Option<AstNodeId> {
+    pub fn set_node_expression(&self, id: ViewNodeId, expression: String) -> Option<()> {
```
Comment (farmaazon) [inline app/gui/src/presenter/graph/state.rs]: We could just return bool, I guess? And what exactly information is returned now? Before we returned the AST ID of the node whose expression needs to be updated. Why we return `()`? It looks like there is no need to update AST in controllers anymore, what sounds strange to me.

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 2]
```diff
@@ -418,7 +418,7 @@ impl Area {
 
             width <- model.label.width.map(|t| t + 2.0 * TEXT_OFFSET);
             frp.output.source.width      <+ width;
-            frp.output.source.expression <+ model.label.content;
+            frp.output.source.expression <+ model.label.content.gate(&edit_mode);
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/component/node/input/area.rs]: I don't like this solution. It makes `expression` output different from the `model.expression.code` if the node happens to be not edited right now. It may cause a host of problems if someone is not aware.

I will do the refactoring you mentioned in the description, especially because it is not a hard one: we could just fire `output.expression` in `init_new_expression` method (where we fire other signals as well).

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/src/presenter/graph/state.rs (hunk 1 above)
Comment (vitvakatu) [reply app/gui/src/presenter/graph/state.rs]: The return value is not used anywhere. So this function can return nothing (if we remove usages of the `?` operator).

File: app/gui/view/graph-editor/src/component/node/input/area.rs (hunk 2 above)
Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node/input/area.rs]: It's not that easy. 
First of all, we can't fire FRP outputs, can we? Outputs don't have the `emit` method, so you need an additional "private" input that would be connected to the output directly.
Secondly, we still need to provide the Presenter with the updates when the user edits the node. So `output.expression` needs to be fired in two places: in `init_new_expression` and in a similar way, on `model.label.content` changes. I will prepare a commit with an alternative implementation.

I guess the main issue is that `model.label` is used for two different activities.
1. Displaying the edited node expression
2. Displaying the `viz_code`
The current implementation relies on the fact that these two are always the same, but it is wrong. Here I filter out "not needed" updates, and I agree that it looks fragile.

File: app/gui/src/presenter/graph/state.rs (hunk 1 above)
Comment (farmaazon) [reply app/gui/src/presenter/graph/state.rs]: Hmm, I would still stay with the old behavior. It's a public method, after all, and all methods of the `ViewChange` returns some information about what needs to be updated in controllers. The fact we ignore it right now seems to be just an incidental thing.

File: app/gui/view/graph-editor/src/component/node/input/area.rs (hunk 2 above)
Comment (farmaazon) [reply app/gui/view/graph-editor/src/component/node/input/area.rs]: Indeed, I missed the fact that `init_new_expression` actually fires the ports _inputs_.

So maybe the most elegant solution is just to have `set_expression` input, and call `set_expression` method when input is fired, and return the new expression to the output (so `set_expression` method is no longer `pub(crate)`)

Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node/input/area.rs]: I implemented this alternative solution by refactoring the `component::node::input::area` module. The refactoring might be hard to follow because huge chunks of code were moved to other places.

File: app/gui/src/presenter/graph/state.rs [hunk 3]
```diff
@@ -444,13 +444,13 @@ impl<'a> ControllerChange<'a> {
...
         let mut nodes = self.nodes.borrow_mut();
         let displayed = nodes.get_mut_or_create(ast_id);
-        tracing::debug!(
-            "Setting node expression from controller: {} -> {}",
-            displayed.expression,
-            new_displayed_expr
-        );
 
         if displayed.expression != new_displayed_expr {
+            tracing::debug!(
```
Comment (wdanilo) [inline app/gui/src/presenter/graph/state.rs]: don't use `tracing::debug` use `debug` instead. Its exported in prelude. Apply everywhere, please.

File: app/gui/view/graph-editor/src/component/node.rs [hunk 4]
```diff
@@ -323,7 +323,7 @@ ensogl::define_endpoints_2! {
         /// Press event. Emitted when user clicks on non-active part of the node, like its
         /// background. In edit mode, the whole node area is considered non-active.
         background_press         (),
-        expression               (enso_text::Rope),
+        expression               (String),
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node.rs]: Why this change? Passing String in FRP is ultra slow. It requires cloning the string on each FRP node. If you really need to pass String, use ImString instead.

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 5]
```diff
@@ -318,250 +320,12 @@ impl Model {
...
-    #[allow(missing_docs)]
-    pub frp:          Frp,
-    pub(crate) model: Rc<Model>,
-}
-
-impl Deref for Area {
-    type Target = Frp;
-    fn deref(&self) -> &Self::Target {
-        &self.frp
-    }
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/area.rs]: Why this code was moved below? Now it is inconsistent - FRP definition is under label "model" while it is not part of Model, its part of Area.

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 6]
```diff
@@ -815,11 +580,11 @@ impl Area {
                         });
                 }
 
-                let index  = node.payload.index;
+                let index = node.payload.index;
                 let length = node.payload.length;
-                let label  = model.label.clone_ref();
+                let label = self.label.clone_ref();
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/area.rs]: I dont like setting FRP from model. we are not doing it anywhere. FPR should be set in Area for model. This is a pattern we are using everywhere. I'd like to keep this pattern. When we will be transitioning to the next FRP implementation this will be required. Can we keep the pattern of setting FRP network in entity that MANAGES model, as it was before, please? If not, lets have a call, please.

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 7]
```diff
@@ -884,26 +649,275 @@ impl Area {
...
+    pub fn get_crumbs_by_id(&self, id: ast::Id) -> Option<Crumbs> {
+        self.model.id_crumbs_map.borrow().get(&id).cloned()
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] All pub functions should have docs.
+    pub fn label(&self) -> &text::Text {
+        &self.model.label
+    }
+
+    /// Set a scene layer for text rendering.
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/area.rs]: if you are changing this code, please fix fixmes.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: app/gui/src/presenter/graph/state.rs (hunk 3 above)
Comment (vitvakatu) [reply app/gui/src/presenter/graph/state.rs]: Done

File: app/gui/view/graph-editor/src/component/node.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node.rs]: I changed it to `ImString`. I can't use `Rope` because, in some code paths, it is not available now.

File: app/gui/view/graph-editor/src/component/node/input/area.rs (hunk 5 above)
Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node/input/area.rs]: Moved the FRP down, closer to the Area

File: app/gui/view/graph-editor/src/component/node/input/area.rs (hunk 6 above)
Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node/input/area.rs]: We're not extending FRP of the Area here, we're extending `port_network` instead.

File: app/gui/view/graph-editor/src/component/node/input/area.rs (hunk 7 above)
Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node/input/area.rs]: All fixes are removed or supplied with a link to the pivotal issue.

**Review by farmaazon — APPROVED** [review]

File: app/gui/view/graph-editor/src/component/node.rs (hunk 4 above)
Comment (wdanilo) [reply app/gui/view/graph-editor/src/component/node.rs]: What od you mean by "not available now"?

**Review by wdanilo — CHANGES_REQUESTED** [review]
1. Some comments are not answered yet:

image

2. It is really hard to review this PR. There are many lines moved around that Git shows as changed and I'm not sure what really changed. This code is a very important one with a LOT of corner cases Even changing the order of some FRP events can affect these corner cases. Checking exactly what changed in this PR would take me hours. Could you please do refactoring of the code (moving it to different places) in another PR / commit, please?

File: app/gui/view/graph-editor/src/component/node.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/view/graph-editor/src/component/node.rs]: So before code changes, `expression` was only produced by `Text::content` FRP output. Now we emit `expression` two times: when the user changes the code (Text::content) and when the code is changed by the Presenter (`set_expression` input). The code behavior is roughly the same, but we don't have `Rope` in the second case. We only have `ImString` from `Expression::code`.

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 8]
```diff
@@ -616,18 +616,24 @@ impl Model {
     }
 
     #[profile(Debug)]
-    fn set_expression(&self, new_expression: impl Into<node::Expression>, area_frp: &FrpEndpoints) {
+    fn set_expression(
+        &self,
+        new_expression: impl Into<node::Expression>,
+        is_editing: bool,
+        area_frp: &FrpEndpoints,
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/area.rs:628]: docs - what is is_editing?

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 9]
```diff
@@ -785,7 +794,19 @@ impl Area {
...
+
+            // === Expression ===
+
+            let frp_endpoints = &frp.output;
+            expression <- frp.input.set_expression.map2(
+                &edit_mode, f!([frp_endpoints, model](expr, is_editing)
+                    model.set_expression(expr, *is_editing, &frp_endpoints)
+                )
+            );
+            frp.output.source.expression <+ expression.map(|e| ImString::new(&e.code));
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/area.rs]: instead of that just do `frp.output.source.expression <+ expression.into();`

File: app/gui/src/presenter/graph/state.rs [hunk 10]
```diff
@@ -437,7 +437,7 @@ impl<'a> ControllerChange<'a> {
         let ast_id = node.main_line.id();
         let new_displayed_expr = node_view::Expression {
             pattern:             node.info.pattern().map(|t| t.repr()),
-            code:                node.info.expression().repr(),
+            code:                ImString::new(node.info.expression().repr()),
```
Comment (wdanilo) [inline app/gui/src/presenter/graph/state.rs]: into

File: app/gui/src/presenter/graph/state.rs [hunk 11]
```diff
@@ -878,7 +878,7 @@ mod tests {
         let view = nodes[0].view;
         let expected_new_expression = view::graph_editor::component::node::Expression {
             pattern:             None,
-            code:                "foo baz".to_string(),
+            code:                ImString::new("foo baz"),
```
Comment (wdanilo) [inline app/gui/src/presenter/graph/state.rs]: into?

File: app/gui/view/debug_scene/interface/src/lib.rs [hunk 12]
```diff
@@ -320,7 +320,13 @@ pub fn expression_mock_string(label: &str) -> Expression {
     let output_span_tree = span_tree::SpanTree::default();
     let input_span_tree = span_tree::SpanTree::new(&ast, &ctx).unwrap();
     let whole_expression_id = default();
-    Expression { pattern, code, whole_expression_id, input_span_tree, output_span_tree }
+    Expression {
+        pattern,
+        code: code.into(),
```
Comment (wdanilo) [inline app/gui/view/debug_scene/interface/src/lib.rs]: can we extract that to var to make it 1 liner?

File: app/gui/view/debug_scene/interface/src/lib.rs [hunk 13]
```diff
@@ -336,7 +342,13 @@ pub fn expression_mock() -> Expression {
     let output_span_tree = span_tree::SpanTree::default();
     let input_span_tree = span_tree::SpanTree::new(&ast, &ctx).unwrap();
     let whole_expression_id = default();
-    Expression { pattern, code, whole_expression_id, input_span_tree, output_span_tree }
+    Expression {
```
Comment (wdanilo) [inline app/gui/view/debug_scene/interface/src/lib.rs]: can we extract that to var to make it 1 liner?

File: app/gui/view/debug_scene/interface/src/lib.rs [hunk 14]
```diff
@@ -376,7 +388,13 @@ pub fn expression_mock2() -> Expression {
         .add_empty_child(36, span_tree::node::InsertionPointType::Append)
         .build();
     let whole_expression_id = default();
-    Expression { pattern, code, whole_expression_id, input_span_tree, output_span_tree }
+    Expression {
+        pattern,
+        code: code.into(),
```
Comment (wdanilo) [inline app/gui/view/debug_scene/interface/src/lib.rs]: can we extract that to var to make it 1 liner?

File: app/gui/view/debug_scene/interface/src/lib.rs [hunk 15]
```diff
@@ -407,5 +425,11 @@ pub fn expression_mock3() -> Expression {
     let output_span_tree = span_tree::SpanTree::new(&ast, &ctx).unwrap(); //span_tree::SpanTree::default();
     let input_span_tree = span_tree::SpanTree::new(&ast, &ctx).unwrap();
     let whole_expression_id = default();
-    Expression { pattern, code, whole_expression_id, input_span_tree, output_span_tree }
+    Expression {
+        pattern,
+        code: code.into(),
```
Comment (wdanilo) [inline app/gui/view/debug_scene/interface/src/lib.rs]: can we extract that to var to make it 1 liner?

**Review by wdanilo — APPROVED** [review]
